sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java

```diff
@@ -85,15 +85,53 @@
 
   /** Information accessible to all methods in this {@code DoFn}. */
   public abstract class Context {
-
     /**
-     * Returns the {@code PipelineOptions} specified with the
-     * {@link org.apache.beam.sdk.runners.PipelineRunner}
-     * invoking this {@code DoFn}.  The {@code PipelineOptions} will
-     * be the default running via {@link DoFnTester}.
+     * Returns the {@code PipelineOptions} specified with the {@link
+     * org.apache.beam.sdk.runners.PipelineRunner} invoking this {@code DoFn}. The {@code
+     * PipelineOptions} will be the default running via {@link DoFnTester}.
      */
     public abstract PipelineOptions getPipelineOptions();
+  }
+
+  /**
+   * Information accessible while within the {@link StartBundle} method.
+   */
+  public abstract class StartBundleContext extends Context {
+  }
 
+  /**
+   * Information accessible while within the {@link FinishBundle} method.
+   */
+  public abstract class FinishBundleContext extends Context {
+    /**
+     * Adds the given element to the main output {@code PCollection} at the given
+     * timestamp in the given window.
+     *
+     * <p>Once passed to {@code output} the element should not be modified in
+     * any way.
+     *
+     * <p><i>Note:</i> A splittable {@link DoFn} is not allowed to output from the
+     * {@link FinishBundle} method.
+     */
+    public abstract void output(OutputT output, Instant timestamp, BoundedWindow window);
```

Comment:
Noting that whenever this context is requested as a parameter, the `DoFn` has to be assumed to be non-window-preserving (analogous to non-key-preserving which is also always assumed).

So we could just say that is always the case and re-add this to `ElementContext`, though the longer term plan is to make these special output receivers instead of context buckets.